samples/dnn/dasiamrpn_tracker.cpp

```diff
@@ -0,0 +1,468 @@
+{
+    Mat baseWindows, HanningWindows;
+
+    createHanningWindow(baseWindows, Size(trackState.scoreSize, trackState.scoreSize), CV_32F);
+    baseWindows = baseWindows.reshape(0, {1, trackState.scoreSize, trackState.scoreSize});
+    HanningWindows = baseWindows.clone();
+    for(int i = 1; i < trackState.anchorNum; i++)
+    {
+        HanningWindows.push_back(baseWindows);
+    }
+
+    return HanningWindows;
+}
+
+Mat generateAnchors(trackerConfig &trackState)
+{
+    int totalStride = trackState.totalStride, scales = trackState.scale, scoreSize = trackState.scoreSize;
+    std::vector<float> ratios = trackState.ratios;
+    std::vector<Rect2f> baseAnchors;
+    int anchorNum = ratios.size();
+    int size = totalStride * totalStride;
+
+    float ori = -(float(scoreSize/2))*float(totalStride);
+
+    for(auto i = 0; i< anchorNum; i++)
+    {
+        int ws = int(sqrt(size/ratios[i]));
+        int hs = int(ws * ratios[i]);
+
+        float wws = float(ws)*scales;
+        float hhs = float(hs)*scales;
+        Rect2f anchor = {0,0,wws,hhs};
+        baseAnchors.push_back(anchor);
+    }
+
+    int anchorIndex[] = {0,0,0,0};
+    const int sizes[] = {4, (int)ratios.size(), scoreSize, scoreSize};
+    Mat anchors(4, sizes, CV_32F);
+
+    for(auto i = 0; i< scoreSize; i++)
+    {
+        for(auto j = 0; j < scoreSize; j++)
+        {
+            for(auto k = 0; k < anchorNum; k++)
+            {
+                anchorIndex[0] = 1, anchorIndex[1] = k, anchorIndex[2] = i, anchorIndex[3] = j;
+                anchors.at<float>(anchorIndex) = ori + totalStride*i;
+
+                anchorIndex[0] = 0;
+                anchors.at<float>(anchorIndex) = ori + totalStride*j;
+
+                anchorIndex[0] = 2;
+                anchors.at<float>(anchorIndex) = baseAnchors[k].width;
+
+                anchorIndex[0] = 3;
+                anchors.at<float>(anchorIndex) = baseAnchors[k].height;
+            }
+        }
+    }
+
+    return anchors;
+}
+
+Mat getSubwindow(Mat &img, const Rect2f &targetBox, float originalSize, Scalar avgChans)
+{
+    Mat zCrop;
+    Size imgSize = img.size();
+    float c = (originalSize+1)/2;
+    float xMin = cvRound(targetBox.x - c);
+    float xMax = xMin + originalSize - 1;
+    float yMin = cvRound(targetBox.y - c);
+    float yMax = yMin + originalSize - 1;
+
+    int leftPad = (int)(fmax(0., -xMin));
+    int topPad = (int)(fmax(0., -yMin));
+    int rightPad = (int)(fmax(0., xMax - imgSize.width +1));
+    int bottomPad = (int)(fmax(0., yMax - imgSize.height +1));
+
+    xMin = xMin + leftPad;
+    xMax = xMax + leftPad;
+    yMax = yMax + topPad;
+    yMin = yMin + topPad;
+
+    if(topPad == 0 && bottomPad ==0 && leftPad == 0 && rightPad == 0)
+    {
+        img(Rect(int(xMin), int(yMin), int(xMax - xMin + 1), int(yMax - yMin + 1))).copyTo(zCrop);
+    } else
+    {
+        copyMakeBorder(img, img, topPad, bottomPad, leftPad, rightPad, BORDER_CONSTANT, avgChans );
```

Comment:
> img, img

Don't use in-place parameters with non-inplace processing.
It has significant performance penalty.

---

`img` is passed as reference here (non-const). Function name (get*) says nothing about modification of passed image.

Reply:
Thanks, fixed.